chromadb/api/client.py

```diff
@@ -0,0 +1,451 @@
+from typing import ClassVar, Dict, Optional, Sequence, TypeVar
+from uuid import UUID
+
+from overrides import override
+from chromadb.api import AdminAPI, ClientAPI, ServerAPI
+from chromadb.api.types import (
+    CollectionMetadata,
+    Documents,
+    EmbeddingFunction,
+    Embeddings,
+    GetResult,
+    IDs,
+    Include,
+    Metadatas,
+    QueryResult,
+)
+from chromadb.config import Settings, System
+from chromadb.telemetry import Telemetry
+from chromadb.telemetry.events import ClientStartEvent
+from chromadb.config import DEFAULT_TENANT, DEFAULT_DATABASE
+from chromadb.api.models.Collection import Collection
+from chromadb.types import Database, Tenant, Where, WhereDocument
+import chromadb.utils.embedding_functions as ef
+
+C = TypeVar("C", "SharedSystemClient", "Client", "AdminClient")
+
+
+class SharedSystemClient:
+    _identifer_to_system: ClassVar[Dict[str, System]] = {}
+    _identifier: str
+
+    # region Initialization
+    def __init__(
+        self,
+        settings: Settings = Settings(),
+    ) -> None:
+        self._identifier = SharedSystemClient._get_identifier_from_settings(settings)
+        SharedSystemClient._create_system_if_not_exists(self._identifier, settings)
+
+    @classmethod
+    def _create_system_if_not_exists(
+        cls, identifier: str, settings: Settings
+    ) -> System:
+        if identifier not in cls._identifer_to_system:
+            new_system = System(settings)
+            cls._identifer_to_system[identifier] = new_system
+
+            new_system.instance(Telemetry)
+            new_system.instance(ServerAPI)
+
+            new_system.start()
+        else:
+            previous_system = cls._identifer_to_system[identifier]
+
+            # For now, the settings must match
+            if previous_system.settings != settings:
+                raise ValueError(
+                    f"An instance of Chroma already exists for {identifier} with different settings"
+                )
+
+        return cls._identifer_to_system[identifier]
+
+    @staticmethod
+    def _get_identifier_from_settings(settings: Settings) -> str:
+        identifier = ""
+        api_impl = settings.chroma_api_impl
+
+        if api_impl is None:
+            raise ValueError("Chroma API implementation must be set in settings")
+        elif api_impl == "chromadb.api.segment.SegmentAPI":
+            if settings.is_persistent:
+                identifier = settings.persist_directory
```

Comment:
somewhat iffy on relying on these as unique id's; why not generate uuids? is it just so the user is able to uniquely address them?

Reply:
the path has to be a unique database, I don't see a way around that